src/app/mesh/qgsmaptooleditmeshframe.cpp

```diff
@@ -2601,6 +2601,16 @@ void QgsMapToolEditMeshFrame::addVertex(
     const QgsPoint layerPoint = mapPointMatch.interpolatedPoint( mCanvas->mapSettings().destinationCrs() );
     zValue = layerPoint.z();
   }
+  else if ( mCurrentEdge.first != -1 && mCurrentEdge.second != -1 ) //we are on a edge -->interpolate the z value
+  {
+    const QVector<int> &edge = edgeVertices( mCurrentEdge );
+    const QgsTriangularMesh &triangularMesh = *mCurrentLayer->triangularMesh();
+    const QgsMeshVertex &v1 = triangularMesh.vertices().at( edge.at( 0 ) );
+    const QgsMeshVertex &v2 = triangularMesh.vertices().at( edge.at( 1 ) );
+    double dist = v1.distance( v2 );
+    double vertDist = QgsGeometryUtils::projectPointOnSegment( QgsPoint( mapPoint ), v1, v2 ).distance( v1 );
```

Comment:
Should we not replace `mapPoint` with the projected one, so that the new point lies exactly on the edge? If we do not "snap" it to the edge, it makes sense to interpolate the Z value from three points as was done previously.

Reply:
@uclaros , good point, the new vertex should be the projected point. Fixed in the last commit